File: solutions/756-pour-water/pour-water.py
class Solution:
    def pourWater(self, heights: List[int], volume: int, k: int) -> List[int]:
        def drop():
            lw_idx, rw_idx = k, k
            l, r = k - 1, k + 1
            while l > -1 and heights[lw_idx] >= heights[l]:
                if heights[lw_idx] > heights[l]: lw_idx = l
                l -= 1
            while r < len(heights) and heights[rw_idx] >= heights[r]:
                if heights[rw_idx] > heights[r]: rw_idx = r
                r += 1
            return lw_idx, rw_idx
        for _ in range(volume):
            lw_idx, rw_idx = drop()
            if heights[lw_idx] < heights[k]: heights[lw_idx] += 1
            else: heights[rw_idx] += 1
        return heights

File: solutions/756-pour-water/pour-water.cpp
class Solution {
public:
    vector<int> pourWater(vector<int>& heights, int V, int K) {
        stack<int> leftFall, rightFall;
        int leftBoundary = K, rightBoundary = K;
        for (int step = 0; step < V; ++step) {
            while (leftBoundary > 0 &&
                   heights[leftBoundary - 1] <= heights[leftBoundary]) {
                --leftBoundary;
                if (heights[leftBoundary] < heights[leftBoundary + 1])
                    leftFall.push(leftBoundary);
            }
            while (rightBoundary < heights.size() - 1 &&
                   heights[rightBoundary + 1] <= heights[rightBoundary]) {
                ++rightBoundary;
                if (heights[rightBoundary] < heights[rightBoundary - 1])
                    rightFall.push(rightBoundary);
            }
            if (!leftFall.empty()) {
                int leftmostFall = leftFall.top();
                ++heights[leftmostFall];
                if (heights[leftmostFall] == heights[leftmostFall + 1])
                    leftFall.pop();
                if (leftmostFall > leftBoundary)
                    leftFall.push(leftmostFall - 1);
            } else if (!rightFall.empty()) {
                int rightmostFall = rightFall.top();
                ++heights[rightmostFall];
                if (heights[rightmostFall] == heights[rightmostFall - 1])
                    rightFall.pop();
                if (rightmostFall < rightBoundary)
                    rightFall.push(rightmostFall + 1);
            } else {
                ++heights[K];
                if (K > leftBoundary)
                    leftFall.push(K - 1);
                if (K < rightBoundary)
                    rightFall.push(K + 1);
            }
        }
        return heights;
    }
};